src/reify.cpp

```diff
@@ -0,0 +1,333 @@
+
+inline void printValue(std::ostream& out, const Sum& s) { out << "sum(" << s.id << "," << s.bound << ")"; }
+
+inline void printValue(std::ostream& out, const Quoted& q) {
+    out.put('"');
+    for (auto c : q.str) {
+        switch (c) {
+            case '\n': out << "\\n"; break;
+            case '\\': out << "\\\\"; break;
+            case '"' : out << "\\\""; break;
+            default  : out.put(c); break;
+        }
+    }
+    out.put('"');
+}
+
+template <typename T, typename... V>
+void printCommaSeparated(std::ostream& out, const T& t, const V&... v) {
+    printValue(out, t);
+    ((out << ",", printValue(out, v)), ...);
+}
+template <typename T>
+struct VectorHash {
+    size_t operator()(const std::vector<T>& vec) const {
+        static_assert(std::is_trivially_copyable_v<T>, "T must be trivially copyable");
+        auto* data = reinterpret_cast<const char*>(vec.data());
+        return std::hash<std::string_view>{}({data, vec.size() * sizeof(T)});
+    }
+};
+
+template <typename T>
+std::vector<T> toVec(std::span<const T> span) {
+    return {span.begin(), span.end()};
+}
+
+}// end unamed namespace
+
+/////////////////////////////////////////////////////////////////////////////////////////
+// Reifier
+/////////////////////////////////////////////////////////////////////////////////////////
+template <typename T>
+using SeqMap = std::unordered_map<std::vector<T>, size_t, VectorHash<T>>;
+
+struct Reifier::StepData {
+    SeqMap<Id_t>      theoryTuples;
+    SeqMap<Id_t>      theoryElementTuples;
+    SeqMap<Lit_t>     litTuples;
+    SeqMap<Atom_t>    atomTuples;
+    SeqMap<WeightLit> weightLitTuples;
+
+    Graph<Atom_t>                        graph;
+    std::unordered_map<Atom_t, uint32_t> nodes;
+
+    void clear() {
+        theoryTuples.clear();
+        theoryElementTuples.clear();
+        litTuples.clear();
+        atomTuples.clear();
+        weightLitTuples.clear();
+        nodes.clear();
+        graph.clear();
+    }
+};
+
+Reifier::Reifier(std::ostream& out, const Options& opts)
+    : out_(out)
+    , calculateSccs_(opts.calculateSccs)
+    , reifyStep_(opts.reifyStep)
+    , stepData_(std::make_unique<StepData>()) {}
+
+Reifier::~Reifier() noexcept = default;
+
+template <typename... T>
+void Reifier::printFact(const char* name, const T&... args) {
+    out_ << name << "(";
+    printCommaSeparated(out_, args...);
+    out_ << ").\n";
+}
+
+template <typename... T>
+void Reifier::printStepFact(const char* name, const T&... args) {
+    if (reifyStep_) {
+        printFact(name, args..., step_);
+    }
+    else {
+        printFact(name, args...);
+    }
+}
+
+template <typename M, typename T>
+auto Reifier::tuple(M& map, const char* name, const T& args) -> size_t {
+    return tuple(map, name, toVec(args));
+}
+
+template <typename M, typename T>
+auto Reifier::tuple(M& map, const char* name, std::vector<T>&& args) -> size_t {
+    std::sort(args.begin(), args.end());
```

Comment:
Consider using `std::ranges::sort` instead - if you follow my suggestion regarding the interface, this would become something like:
```cpp
template <typename M, typename T>
auto Reifier::tuple(M& map, const char* name, std::span<T> args) -> size_t {
    auto owned = toVec(args);
    std::ranges::sort(owned);
    auto [it, inserted] = map.emplace(std::move(owned), map.size());
   ...
```